// @react-hookz/web/cjs/util/misc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.basicDepsComparator = exports.yieldFalse = exports.yieldTrue = exports.hasOwnProperty = exports.off = exports.on = void 0;
function on(obj) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    if (obj && obj.addEventListener) {
        obj.addEventListener.apply(obj, args);
    }
}
exports.on = on;
function off(obj) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    if (obj && obj.removeEventListener) {
        obj.removeEventListener.apply(obj, args);
    }
}
exports.off = off;
var hasOwnProperty = function (obj, property) { return Object.prototype.hasOwnProperty.call(obj, property); };
exports.hasOwnProperty = hasOwnProperty;
var yieldTrue = function () { return true; };
exports.yieldTrue = yieldTrue;
var yieldFalse = function () { return false; };
exports.yieldFalse = yieldFalse;
var basicDepsComparator = function (d1, d2) {
    if (d1 === d2)
        return true;
    if (d1.length !== d2.length)
        return false;
    for (var i = 0; i < d1.length; i++) {
        if (d1[i] !== d2[i]) {
            return false;
        }
    }
    return true;
};
exports.basicDepsComparator = basicDepsComparator;
